app/mltraining_client.go

```diff
@@ -0,0 +1,323 @@
+	TrainingStatusInProgress
+	// TrainingStatusCompleted respresents a completed training job.
+	TrainingStatusCompleted
+	// TrainingStatusFailed respresents a failed training job.
+	TrainingStatusFailed
+	// TrainingStatusCanceled respresents a canceled training job.
+	TrainingStatusCanceled
+	// TrainingStatusCanceling respresents a training job that is being canceled.
+	TrainingStatusCanceling
+)
+
+// ErrorDetail contains an arbitrary serialized protocol buffer message along with a URL that describes the type of serialized message.
+type ErrorDetail struct {
+	TypeURL string
+	Value   []byte
+}
+
+// ErrorStatus contains an error's code, message, and details.
+type ErrorStatus struct {
+	Code    int
+	Message string
+	Details []*ErrorDetail
+}
+
+// TrainingJobMetadata contains the metadata for a training job.
+type TrainingJobMetadata struct {
+	ID                  string
+	DatasetID           string
+	OrganizationID      string
+	ModelName           string
+	ModelVersion        string
+	ModelType           ModelType
+	ModelFramework      ModelFramework
+	IsCustomJob         bool
+	RegistryItemID      string
+	RegistryItemVersion string
+	Status              TrainingStatus
+	ErrorStatus         *ErrorStatus
+	CreatedOn           *time.Time
+	LastModified        *time.Time
+	TrainingStarted     *time.Time
+	TrainingEnded       *time.Time
+	SyncedModelID       string
+	Tags                []string
+}
+
+// GetTrainingJobLogsOptions contains optional parameters for GetTrainingJobLogs.
+type GetTrainingJobLogsOptions struct {
+	PageToken *string
+}
+
+// TrainingJobLogEntry is a log entry from a training job.
+type TrainingJobLogEntry struct {
+	Level   string
+	Time    *time.Time
+	Message string
+}
+
+// MLTrainingClient is a gRPC client for method calls to the ML Training API.
+type MLTrainingClient struct {
+	client pb.MLTrainingServiceClient
+}
+
+func newMLTrainingClient(conn rpc.ClientConn) *MLTrainingClient {
+	return &MLTrainingClient{client: pb.NewMLTrainingServiceClient(conn)}
+}
+
+// SubmitTrainingJob submits a training job request and returns its ID.
+func (c *MLTrainingClient) SubmitTrainingJob(
+	ctx context.Context, datasetID, organizationID, modelName, modelVersion string, modelType ModelType, tags []string,
+) (string, error) {
+	resp, err := c.client.SubmitTrainingJob(ctx, &pb.SubmitTrainingJobRequest{
+		DatasetId:      datasetID,
+		OrganizationId: organizationID,
+		ModelName:      modelName,
+		ModelVersion:   modelVersion,
+		ModelType:      modelTypeToProto(modelType),
+		Tags:           tags,
+	})
+	if err != nil {
+		return "", err
+	}
+	return resp.Id, nil
+}
+
+// SubmitCustomTrainingJob submits a custom training job request and returns its ID.
+func (c *MLTrainingClient) SubmitCustomTrainingJob(
+	ctx context.Context,
+	datasetID,
+	registryItemID,
+	registryItemVersion,
+	organizationID,
+	modelName,
+	modelVersion string,
+	arguments map[string]string,
```

Comment:
hm... I'm a little worried about the number of arguments here. I realize we have the pattern in python of listing all the arguments individually and there is some value of maintaining consistency with that, but also python supports named arguments which makes it easier to verify that you're passing the right value in the right place. We also just ask for the bare arguments in flutter unfortunately, and it's probably too late to change there. 

I wonder if it would be useful to have a `customTrainingJobArgs` struct that contained all the necessary information? That way it would be a lot easier for users to call this with confidence that they've assigned the args correctly. OTOH, a struct will have default constructors for undefined fields, which makes it easier for a user to just accidentally not pass a field at all... 

both approaches seem defensible imo. I personally prefer the idea of a struct (slightly), but I do think both create opportunities for users to accidentally screw up. What do you think? I'm happy to go with your take if you have one (or see if @njooma has a thought, or if @benjirewis is aware of some golang idiom that we should prefer here). I think the important thing is just that we be intentional with this choice, since it'll be difficult to roll back once it's in place.

Reply:
(this same comment can apply to the `SubmitTrainingJob` method above).

Reply:
I would be okay with whatever is the norm in Golang.
Personally, I like to have the parameters in the function as I have a pop-up show up to tell me what parameter to name next. Having a struct would actually be more reading/confusing for me as a user, although I'm sure not everyone has this setup. Having a user make a struct means that the definitions are more deliberate, even though ironically there's now a chance of completely missing a field.
As for the possibility of undefined fields, in these two functions in particular, the method can check all the fields and return an error if it's empty.

Side note: If going this way, I'd want to do this for all the functions with many parameters. In those functions, however, there would be instances where certain fields can/will be left undefined on purpose. It wouldn't be as easy to check if a field was missed, so I guess we stay with just doing it here.

Reply:
I added `SubmitTrainingJobArgs` and a new `isValid` function. I kept some of the arguments out since some of them (`tags` and `arguments`) can be left empty on purpose. Let me know if I should make a separate args struct for custom that includes the registry item information.